### PR TITLE
Refactor gem with updated apis

### PR DESCRIPTION
**Pull request by briancain**
This is a pull request for updating the `vagrant_cloud` gem. Some of these updates include:

- Updated to match the latest Vagrant Cloud API functionality
- Refactored to have a `Client` class in charge of making API requests separate from the `Account` class
- Allows for making _one-off_ API requests for when making and holding onto an entirely new class object to make Vagrant Cloud API requests does not make sense

These changes _should_ be backwards compatible, and should not break existing clients at all.

### CODE REVIEW

**Comment by briancain** [comment]
Hmm, looks like I need to update a function or two for older ruby versions.

File: lib/vagrant_cloud/box.rb [hunk 1]
```diff
@@ -7,12 +7,94 @@ class Box
...
+      short_description = @short_description if !short_description
+      description = @description if !description
+
+      params = {
+        name: box_name,
+        username: org,
+        is_private: is_private,
+        short_description: short_description,
+        description: description,
+      }.delete_if { |k, v| v.nil? }
```
Comment (chrisroberts) [inline lib/vagrant_cloud/box.rb]: `delete_if { |_, v|`

File: lib/vagrant_cloud/provider.rb [hunk 2]
```diff
@@ -23,29 +32,82 @@ def download_url
...
+    # @param [String] box_name
+    # @param [String] version_number
+    def create_provider(name=nil, url = nil, username=nil, box_name=nil, version_number=nil)
+      name = @name if !name
+      url = @url if !url
+      username = @username if !username
+      box_name = @box_name if !box_name
+      version_number = @version if !version
+
+      params = { name: name, url: url }.delete_if { |_k, v| v.nil? }
```
Comment (chrisroberts) [inline lib/vagrant_cloud/provider.rb]: `delete_if { |_, v|`

File: lib/vagrant_cloud/search.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    # @return [Hash]
+    def search(query=nil, provider=nil, sort=nil, order=nil, limit=nil, page=nil)
+      params = {
+        q: query,
+        provider: provider,
+        sort: sort,
+        order: order,
+        limit: limit,
+        page: page
+      }.delete_if { |k, v| v.nil? }
```
Comment (chrisroberts) [inline lib/vagrant_cloud/search.rb]: `delete_if { |_, v|`

File: lib/vagrant_cloud/version.rb [hunk 4]
```diff
@@ -73,7 +137,7 @@ def get_provider(name, data = nil)
     # @return [Provider]
     def create_provider(name, url = nil)
       params = { name: name, url: url }.delete_if { |_k, v| v.nil? }
```
Comment (chrisroberts) [inline lib/vagrant_cloud/version.rb]: `delete_if { |_, v|`

File: vagrant_cloud.gemspec [hunk 5]
```diff
@@ -13,7 +13,8 @@ Gem::Specification.new do |s|
   s.executables << 'vagrant_cloud'
 
   s.add_runtime_dependency 'rest-client', '~> 2.0.2'
-  s.add_runtime_dependency 'thor', '~> 0.19.4'
+  #s.add_runtime_dependency 'thor', '~> 0.19.4'
```
Comment (chrisroberts) [inline vagrant_cloud.gemspec]: Delete this commented dependency line

File: vagrant_cloud.gemspec [hunk 6]
```diff
@@ -13,7 +13,8 @@ Gem::Specification.new do |s|
   s.executables << 'vagrant_cloud'
```
Comment (chrisroberts) [inline vagrant_cloud.gemspec]: Remove this executables entry

File: vagrant_cloud.gemspec [hunk 7]
```diff
@@ -13,7 +13,8 @@ Gem::Specification.new do |s|
...
 
   s.add_runtime_dependency 'rest-client', '~> 2.0.2'
-  s.add_runtime_dependency 'thor', '~> 0.19.4'
+  #s.add_runtime_dependency 'thor', '~> 0.19.4'
+  s.add_runtime_dependency 'json', '~> 2.1.0'
 
   s.add_development_dependency 'rake', '~> 10.4'
   s.add_development_dependency 'rspec', '~> 3.0'
```
Comment (chrisroberts) [inline vagrant_cloud.gemspec]: Should add a post install message to notify users of the library update and CLI removal

**Pull request merged**